### ISSUE TITLE
What if I implement a focal loss?

### ISSUE THREAD

**Issue by DonghoonPark12** [luuuyi/RefineDet.PyTorch#26]
If I want to utilize focal loss. where should I change in this code? 

I guess change in `refinedet_multibox_loss.py` `loss_c = F.cross_entropy(conf_p, targets_weighted, reduction='sum')` to focal loss.

**Comment by luuuyi** [comment]
@DonghunP I think you should change the softmax compute in this file, because in refinedet, it chooses the most harder negative examples depend on 3:1 ratio. If you change to focal loss, you should utilize all the negative sample.

**Comment by DonghoonPark12** [comment]
@luuuyi I changed to 
```python
# num_neg = torch.clamp(self.negpos_ratio*num_pos, max=pos.size(1)-1)
# neg = idx_rank < num_neg.expand_as(idx_rank)

# neg = idx_rank # I changed to below. Because, It occur error.
neg = idx_rank.type(torch.uint8)
```
as you said(I also agree) and,
```python
loss_c = F.cross_entropy(conf_p, targets_weighted, reduction='sum')
# Refer : https://www.kaggle.com/c/tgs-salt-identification-challenge/discussion/65938
pt = torch.exp(-loss_c) # By multiply exp, -log(p) convert to p if y==1, -log(1-p) convert to 1-p if y!=1.
F_loss = self.alpha * (1 - pt) ** self.gamma * loss_c #self.apha = 0.75, self.gamma=2.0

N = num_pos.data.sum().float()
loss_l /= N
F_loss /= N
return loss_l, F_loss 
```

in refinedet_multibox_loss.py

What do you think?

---
During training, loss becomes `inf`.
What is the meaning of `change the softmax compute in this file` ?

**Comment by luuuyi** [comment]
@DonghunP I mean change the softmax compute to focal loss compute

**Comment by DonghoonPark12** [comment]
@luuuyi Thanks. I did.
